src/relay/transforms/pattern_utils.h

```diff
@@ -120,6 +120,21 @@ inline Expr InferType(const Expr& expr) {
   }
 }
 
+/*!
+ * \brief Try to apply constant folding over expr:
+ *
+ * Do constant_fold over each node in expr
+ *
+ * \param expr The IR expression
+ * \return resulting expr
+ */
+inline Expr FoldConstantExpr(const Expr& expr) {
+  IRModule const_mod = IRModule::FromExpr(expr);
+  const_mod = transform::FoldConstant()(const_mod);
+  GlobalVar const_main = const_mod->GetGlobalVar("main");
+  return Downcast<Function>(const_mod->functions[const_main])->body;
+}
```

Comment:
There is already https://github.com/apache/tvm/blob/78142ad2e4ce2c5ca096f7e80f4f4f62ca9c2bd5/src/relay/transforms/fold_constant.cc#L429. You can probably just use that.

And it's odd to add this function in `pattern_utils.h`.

Reply:
> You can probably just use that.
Thanks, I did see that one, but (1). currently it doesn't seem to be declared in any header, and (2). the extra `mod` argument is not ergonomic for our use case here.

> And it's odd to add this function in pattern_utils.h.
I agree, I did this for consistency because we had [InferType](https://github.com/apache/tvm/blob/main/src/relay/transforms/pattern_utils.h#L113-L121) right above from #13275.

I think they should probably both be moved to either `pass_utils.h` or a new `transform_utils.h`.

Let me know what you think, I'll update the PR accordingly:).

Reply:
Okay how about this:

* Introduce `fold_constant.h` and declare `FoldConstantExpr(const Expr& expr)` and `FoldConstantExpr(const Expr& expr, const IRModule& mod, bool fold_qnn)`.
* Implement the former one in terms of the latter.
* Replace all duplicated definition of `FoldConstantExpr` with these ones.

Reply:
I like that, done!

Caught another duplication in `relay/quantize/realize.cc`. Let me know if it needs more tweaking.